### PR TITLE
Integration comet dev

### PR DESCRIPTION
**Pull request by DN6**

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -52,91 +52,127 @@
...
+    check_suffix(weights, ".pt")  # check weights
+    pretrained = weights.endswith(".pt")
     if pretrained:
         with torch_distributed_zero_first(LOCAL_RANK):
             weights = attempt_download(weights)  # download if not found locally
-        ckpt = torch.load(weights, map_location='cpu')  # load checkpoint to CPU to avoid CUDA memory leak
-        model = Model(cfg or ckpt['model'].yaml, ch=3, nc=nc, anchors=hyp.get('anchors')).to(device)  # create
-        exclude = ['anchor'] if (cfg or hyp.get('anchors')) and not resume else []  # exclude keys
-        csd = ckpt['model'].float().state_dict()  # checkpoint state_dict as FP32
+            ckpt = torch.load(weights, map_location="cpu")  # load checkpoint to CPU to avoid CUDA memory leak
```
Comment (Lothiraldan) [inline train.py]: Is this change intended?

File: utils/loggers/__init__.py [hunk 2]
```diff
@@ -37,14 +47,32 @@
 try:
     import clearml
 
-    assert hasattr(clearml, '__version__')  # verify package import not local dir
+    assert hasattr(clearml, "__version__")  # verify package import not local dir
 except (ImportError, AssertionError):
     clearml = None
 
+try:
+    assert RANK in {0, -1}
```
Comment (Lothiraldan) [inline utils/loggers/__init__.py]: Let's not use `assert` as it can be disabled, let's check the value with:
```
if RANK not in [0, -1]:
  comet_ml = None
else:
  import comet_ml
```

File: utils/loggers/__init__.py [hunk 3]
```diff
@@ -53,95 +81,159 @@ def __init__(self, save_dir=None, weights=None, opt=None, hyp=None, logger=None,
...
         else:
             self.clearml = None
 
+        # Comet
+        if comet_ml and "comet" in self.include:
+            try:
+                if isinstance(self.opt.resume, str) and self.opt.resume.startswith(
+                    "comet://"
+                ):
+                    run_id = self.opt.resume.split("/")[-1]
```
Comment (Lothiraldan) [inline utils/loggers/__init__.py]: What are the expected format of the resume? If it's a URL (like `comet://experiment-id/`), it might be cleaner to use https://docs.python.org/3/library/urllib.parse.html?highlight=urlparse#urllib.parse.urlparse

File: utils/loggers/__init__.py [hunk 4]
```diff
@@ -240,21 +380,26 @@ class GenericLogger:
         include:         loggers to include
     """
 
-    def __init__(self, opt, console_logger, include=('tb', 'wandb')):
+    def __init__(self, opt, console_logger, include=("tb", "wandb")):
```
Comment (Lothiraldan) [inline utils/loggers/__init__.py:337]: When is the `GenericLogger` used? Should we update it to call Comet?

File: utils/loggers/comet/__init__.py [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+from utils.dataloaders import img2label_paths
+from utils.general import scale_coords, xywh2xyxy
+from utils.metrics import ConfusionMatrix, box_iou
+
+LOGGER = logging.getLogger(__name__)
+
+COMET_PREFIX = "comet://"
+
+# Project Configuration
+COMET_PROJECT_NAME = os.getenv("COMET_PROJECT_NAME", "yolov5")
```
Comment (Lothiraldan) [inline utils/loggers/comet/__init__.py]: A lot of configuration keys can also be configured in a configuration file. It would be better to use something like:
```
config = comet_ml.config.get_config()
COMET_PROJECT_NAME = config.get_string(USER_VALUE, "comet.project_name", default="yolov5")
```

File: utils/loggers/comet/__init__.py [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+                self.experiment.log_other("Created from", "YOLOv5")
+
+                if not isinstance(self.experiment, comet_ml.OfflineExperiment):
+                    workspace, project_name, experiment_id = self.experiment.url.split(
+                        "/"
+                    )[-3:]
+                    self.experiment.log_other(
+                        "Run Path",
+                        f"{workspace}/{project_name}/{experiment_id}",
+                    )
```
Comment (Lothiraldan) [inline utils/loggers/comet/__init__.py:114]: Why do we need to log the Experiment Run Path? If the experiment is copied to another project / workspace, that value is gonna be outdated

File: utils/loggers/comet/utils.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        if (int(checkpoint_step) == asset["step"]) and (
+            checkpoint_filename == asset["fileName"]
+        ):
+            asset_id = asset["assetId"]
+            asset_filename = asset["fileName"]
+            break
+
+    try:
+        model_binary = experiment.get_asset(
+            asset_id, return_type="binary", stream=False
```
Comment (Lothiraldan) [inline utils/loggers/comet/utils.py]: How big would the model checkpoint will be? With `stream=False` I think it will have to put the file content in memory

**Review by Lothiraldan — COMMENTED** [review]
Could you clarify the user-flow about Artifact, when would an artifact be used as input and what is logged as an Artifact as output? If have seen somewhere in the code that if you are resuming, I we won't log a new artifact so I'm bit in the dark regarding the user-flow.

When we are logging the Confusion Matrix, are we logging images as well?

For the checkpoint file, are they logged in the same Experiment Model with the same name but different steps?

File: utils/loggers/__init__.py (hunk 4 above)
Comment (DN6) [reply utils/loggers/__init__.py:337]: Hmm I'm not sure about this one. Will have to check.

File: utils/loggers/comet/__init__.py (hunk 6 above)
Comment (DN6) [reply utils/loggers/comet/__init__.py:114]: This is logged as a way to conveniently resume an experiment or use weights from an experiment 

Resuming
```
python train.py --resume "comet://<run path>"
``` 

Using Weights
```
python train.py --weights "comet://<run path>"
```

File: utils/loggers/comet/utils.py (hunk 7 above)
Comment (DN6) [reply utils/loggers/comet/utils.py:36]: Good catch. The weights could go up to a GB.

File: train.py (hunk 1 above)
Comment (DN6) [reply train.py]: Hmm this one may have come out of a merge conflict resolution. Let me double check.
